contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,857 @@
+    requiredProp: PropTypes.string.isRequired,
+    optionalPropWithDefaultValue: PropTypes.number,
+    optionalProp: PropTypes.bool,
+}
+
+const defaultProps = {
+    optionalPropWithDefaultValue: 42,
+};
+
+function Foo(props) {
+    // your component's code
+}
+
+Foo.propTypes = propTypes;
+Foo.defaultProps = defaultProps;
+
+export default Foo;
+
+// After
+type Props = {
+    requiredProp: string;
+    optionalPropWithDefaultValue?: number;
+    optionalProp?: boolean;
+}
+
+function Foo({ requiredProp, optionalPropWithDefaultValue = 42, optionalProp }: Props) {
+    // your component's code
+}
+
+export default Foo;
+```
+
+### Style props
+
+When converting or typing style props, use `StyleProp<T>` type where `T` is the type of styles related to the component your prop is going to apply.
+
+- If your style prop is going to be applied to a `View`, use `StyleProps<ViewStyle>`.
+- If your style prop is going to be applied to a `Text`, use `StyleProps<TextStyle>`.
+- If your style prop is going to be applied to a `Image`, use `StyleProps<ImageStyle>`.
+
+```ts
+// Before
+const propTypes = {
+    containerStyle: PropTypes.arrayOf(PropTypes.object),
+    textStyle: PropTypes.arrayOf(PropTypes.object),
+    imageStyle: PropTypes.arrayOf(PropTypes.object),
+};
+
+// After
+import {StyleProp, ViewStyle, TextStyle, ImageStyle} from 'react-native'
+
+type Props = {
+    containerStyle: StyleProp<ViewStyle>;
+    textStyle: StyleProp<TextStyle>;
+
+    // If the styling is optional, mark the prop as optional too.
+    imageStyle?: StyleProp<ImageStyle>;
+};
+```
+
+### Animated styles
+
+If you need to pass style props to an Animated component, use `Animated.WithAnimatedValue<T>` to augment and add support for animated values in your prop.
+
+```tsx
+import {useRef} from 'react';
+import {Animated, StyleProp, ViewStyle} from 'react-native';
+
+type Props = {
+    style?: Animated.WithAnimatedValue<StyleProp<ViewStyle>>;
+};
+
+function Component(props: Props) {
+    return <Animated.View style={props.style} />;
+}
+
+function App() {
+    const anim = useRef(new Animated.Value(0)).current;
+    return <Component style={{opacity: anim.interpolate({...})}} />;
+}
+```
+
+### Refs
+
+Use `React.forwardRef` to pass refs between parent and child components. Utilize `React.forwardRef` generic parameters to provide types. The first parameter specifies the **ref's element type** - common components such as `View`, `TextInput` and `Text`. The second parameter is the **props type of the wrapped component**.
+
+```tsx
+const ChildComponent = React.forwardRef<View, ViewProps>((props, ref) => (
+    <View {...props} ref={ref}>
+        <Text>Component</Text>
+    </View>
+));
+
+function ParentComponent() {
+  const viewRef = React.useRef<View>(null);
+
+  return (
+    <View>
+      <ChildComponent ref={viewRef} {...someViewProps} />
+    </View>
+  );
+};
+```
+
+### Children and render props
+
+To type children props, use `React.ReactNode`.
```

Comment:
Question: did we decide against `PropsWithChildren`, or should we add an example here?

I generally prefer it over `ReactNode` because it makes the type easier to read (one less property), although I'm not sure how well it works with render props

Reply:
I don't have a strong opinion on this. Both `PropsWithChildren` and typing children with `ReactNode` are good approaches. I would leave it as is it written in the guidelines just because it is easier for non-Typescript users as you don't need to use generics in that case.